### PR TITLE
lazily calls some fns

### PR DESCRIPTION
**Pull request by klensy**
Replaced some fn's with it's lazy variants.

### CODE REVIEW

File: compiler/rustc_middle/src/ty/query/on_disk_cache.rs [hunk 1]
```diff
@@ -525,7 +525,7 @@ impl<'sess> OnDiskCache<'sess> {
     ) {
         let mut current_diagnostics = self.current_diagnostics.borrow_mut();
 
-        let x = current_diagnostics.entry(dep_node_index).or_insert(Vec::new());
+        let x = current_diagnostics.entry(dep_node_index).or_insert_with(Vec::new);
```
Comment (jonas-schievink) [inline compiler/rustc_middle/src/ty/query/on_disk_cache.rs:528]:
```suggestion
        let x = current_diagnostics.entry(dep_node_index).or_default();
```

File: compiler/rustc_mir/src/borrow_check/diagnostics/outlives_suggestion.rs [hunk 2]
```diff
@@ -157,7 +157,7 @@ impl OutlivesSuggestionBuilder {
         debug!("Collected {:?}: {:?}", fr, outlived_fr);
 
         // Add to set of constraints for final help note.
-        self.constraints_to_add.entry(fr).or_insert(Vec::new()).push(outlived_fr);
+        self.constraints_to_add.entry(fr).or_insert_with(Vec::new).push(outlived_fr);
```
Comment (jonas-schievink) [inline compiler/rustc_mir/src/borrow_check/diagnostics/outlives_suggestion.rs:160]:
```suggestion
        self.constraints_to_add.entry(fr).or_default().push(outlived_fr);
```

File: compiler/rustc_resolve/src/late.rs [hunk 3]
```diff
@@ -2327,7 +2327,7 @@ impl<'a: 'ast, 'b, 'ast> LateResolutionVisitor<'a, 'b, 'ast> {
 
             ExprKind::Call(ref callee, ref arguments) => {
                 self.resolve_expr(callee, Some(expr));
-                let const_args = self.r.legacy_const_generic_args(callee).unwrap_or(Vec::new());
+                let const_args = self.r.legacy_const_generic_args(callee).unwrap_or_else(Vec::new);
```
Comment (jonas-schievink) [inline compiler/rustc_resolve/src/late.rs:2330]:
```suggestion
                let const_args = self.r.legacy_const_generic_args(callee).unwrap_or_default();
```

**Comment by the8472** [comment]
Note that Vec::new doesn't allocate and generaly has no side-effects since it's const. So it doesn't really make a difference whether it's called lazily or not.

File: compiler/rustc_middle/src/ty/query/on_disk_cache.rs (hunk 1 above)
Comment (klensy) [reply compiler/rustc_middle/src/ty/query/on_disk_cache.rs]: Applied this one and others, thanks, looks more clear.

**Comment by klensy** [comment]
@the8472 I understand, but it looks little more accurate, even better with `default` methods.

**Pull request merged**